### DISCUSSION TITLE
getting immediate state for store subscribers

### DISCUSSION THREAD

**Discussion by rotty3000** [microsoft/redux-micro-frontend#15] [Q&A]
Is there a way in the subscriber only scenario to "get" the state immediately? 

Otherwise I have to wait for a store event to occur. 

Currently to work around this I have to implement a global `ping` action and fire that in order for my subscriber function to get called.

Is there a better way?

**Comment by rotty3000** [comment]
Another issue in the subscribe only scenario is that if the store doesn't exist yet I get an error. I'm wondering if there's a way to handle this as an "subscribe on demand" approach.

**Reply by rotty3000** [reply]
On possible approach could be to use `GlobalStore.Get().SubscribeToGlobalState(...)`

**Comment by tmonikat** [comment]
> Is there a way in the subscriber only scenario to "get" the state immediately?
> 
> Otherwise I have to wait for a store event to occur.
> 
**> Currently to work around this I have to implement a global `ping` action and fire that in order for my subscriber function to get called.**
> 
> Is there a better way?

How do you implement a global ping? 
I am also having an issue where the Subscriber MFE events are not getting fired.

In my subscribed mfe.

```
 const globalContext = useContext(GlobalContext);

 if (globalContext ) {
        console.log("Registering CustomerAPP");
        globalContext .RegisterStore("CustomerApp", store);
        
        globalContext .RegisterGlobalActions("CustomerApp", [custFetchConstants.BLOG_FETCH_REQUEST, 
  custFetchConstants.BLOG_FETCH_SUCCESS,custFetchConstants.BLOG_FETCH_ERROR]);
    }

```
On local reducer action, I notice the Subscribe block is executed 3 times on first time page load. 
After first time page load and action, for the next 'createCustCall' the Subscribe even doesnt trigger and GlobalStore still contains stale data.

```
 const createCustCall =() => {
        dispatch(createCustomer({ id, name}));
        globalStore && globalStore.SubscribeToGlobalState("CustomerApp", state => {
            console.log("After create CustomerApp state=",state);
        });

    }
```

How do you invoke Global ping here,  to trigger subscribe action?

**Reply by rotty3000** [reply]
> How do you implement a global ping?

naively.. as follows:
```js
function navigationReducer(state: any = initialState, action: any): any {
	switch (action.type) {
		case ...
		default:
			return state;
	}
}

let globalStore:GlobalStore = GlobalStore.Get();

let navStore:Redux.Store = createStore(
	navigationReducer,
	...
);

globalStore.RegisterStore(
	'foo',
	navStore
);

globalStore.RegisterGlobalActions(
	'foo',
	[... , 'navigation/ping']
);
```
and later, but eagerly:
```js
GlobalStore.Get().DispatchAction('foo', <any> {type: 'navigation/ping'});
```

**Reply by tmonikat** [reply]
Thanks for your suggestion @rotty3000 .
I was able resolve the issue of immediate state update by updating the signature of RegisterStore() as follows:

```
const { globalStore } = props;
    if (globalStore) {
        **_globalStore.RegisterStore("CustomerApp", store, [], true);_**
        console.log("Register CustomerApp state=",globalStore);
    }
```
The change follows this:
```
    /**
     * Summary: Registers an isolated app store
     *
     * @access public
     *
     * @param {string} appName Name of the App.
     * @param {Store} store Instance of the store.
     * @param {boolean} shouldReplace Flag to indicate if the an already registered store needs to be replaced.
     */
      RegisterStore(appName: string, store: Store, globalActions?: Array<string>, shouldReplaceExistingStore?: boolean): void;
```
The last boolean value to replace the store, did the magic.
After making the above change, I can see the state update events are triggered immediately for all local state update.

However, the second MFE  still cant listen to the events.
```
const { globalStore } = useContext(AppContext);
 useEffect(() => {
       if (globalStore && globalStore.GetPartnerState("CustomerApp")) {
           //this works
            setBlogs(globalStore.GetPartnerState("CustomerApp").customer.data);
           //this doesn't trigger as it is not on current page
            globalStore.Subscribe('CustomerApp', state => {
               //this is NEVER printed
                console.log("From CustomerApp state=",state);
                setBlogs(state.customer.data);
            });
       }
    }, [globalStore]);
```
Any suggestions to improve on listening to events.

**Comment by rishabh01071967** [comment]
if I am doing dispatch global action in one micro front end application. so in the another micro front end application do we have to dispacth some action there as well as to fetch the state as  subscribetopartnerstate or subscribetoglobalState not work till i do some dispatch.
